### PR TITLE
8432 Coop IA - upload rules step

### PR DESCRIPTION
**Pull request by argush3**
*Issue #:* /bcgov/entity#8432

*Description of changes:*

* Implementation of upload rules step for COOP IA

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/assets/styles/theme.scss
Comment (argush3) [reply src/assets/styles/theme.scss]: Removed and `$BCgovFineInputBG` and updated to use `$gray1` instead.

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: done

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: done

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: removed line break elements and added top margin to next element

File: src/store/mutations/mutations-model.ts
Comment (argush3) [reply src/store/mutations/mutations-model.ts]: removed

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue:129]: added whitespace

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: updated to use `10 * 1024`

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: updated to use `:inputFile="{...uploadRulesDoc}"`

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: updated to use camel case

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue:185]: updated calling function(`fileSelected`) to be async and added await to `uploadPendingDocsToStorage` invocation

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: done

File: src/components/common/FileUploadPreview.vue
Comment (argush3) [reply src/components/common/FileUploadPreview.vue]: fixed

File: src/components/common/FileUploadPreview.vue
Comment (argush3) [reply src/components/common/FileUploadPreview.vue]: updated to use `$app-blue`

File: src/interfaces/stepper-interfaces/CreateRules/create-rules-interface.ts
Comment (argush3) [reply src/interfaces/stepper-interfaces/CreateRules/create-rules-interface.ts]: removed commas

File: src/store/getters/state-getters.ts
Comment (argush3) [reply src/store/getters/state-getters.ts]: updated to use `CreateRulesIF` return type

File: src/components/CreateRules/UploadRules.vue
Comment (severinbeauvais) [reply src/components/CreateRules/UploadRules.vue]: I thought it might be easier to read with a blank line between the header the section, but YMMV.

File: src/components/CreateRules/UploadRules.vue
Comment (severinbeauvais) [reply src/components/CreateRules/UploadRules.vue]: Oops, I didn't mean "destructure".

I mean to change
`v-bind:input-file="uploadRulesDoc"`
to
`:input-file="uploadRulesDoc"`.

(The ... was me trying to avoid typing what followed v-bind ;) )

File: src/components/CreateRules/UploadRules.vue
Comment (severinbeauvais) [reply src/components/CreateRules/UploadRules.vue:233]: Ditto for help-section-label-top and help-section-label: it's sometimes/usually easier to use the Vuetify spacing helpers (eg, mt-4) for layout purposes, in the template, instead of using a classname that a reviewer has to look up to see what the layout is.
Again, YMMV. This is not a blocking change for this PR.

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue]: ah...haha.  ok, will fix.

File: src/resources/cooperative.ts
Comment (cameron-freshworks) [reply src/resources/cooperative.ts:64]: If it's easier without using the resource file and there is no future configuration required, it probably makes sense to go without! Your call though 👍

File: src/interfaces/stepper-interfaces/CreateRules/rules-doc-interface.ts
Comment (argush3) [reply src/interfaces/stepper-interfaces/CreateRules/rules-doc-interface.ts:3]: A number should be fine here.  Its value is retrieve from the `File` object.  

>A number that represents the number of milliseconds since the Unix epoch.

https://developer.mozilla.org/en-US/docs/Web/API/File/lastModified

I did remove the commas from the interface however.

File: src/resources/cooperative.ts
Comment (argush3) [reply src/resources/cooperative.ts:87]: updated to support new window

File: src/interfaces/stepper-interfaces/CreateRules/create-rules-resource-interface.ts
Comment (argush3) [reply src/interfaces/stepper-interfaces/CreateRules/create-rules-resource-interface.ts]: updated to use `any`

File: src/resources/cooperative.ts
Comment (argush3) [reply src/resources/cooperative.ts:64]: ended up sticking with the resource file.

Comment (cameron-freshworks) [reply src/resources/cooperative.ts:64]: Sounds good!

**Review by cameron-freshworks — APPROVED** [review]
Looks good to me, when the rest of the comments are resolved! 👍

File: src/mixins/filing-template-mixin.ts
Comment (vysakh-menon-aot) [inline src/mixins/filing-template-mixin.ts:100]: We don't need rules and memorandum in incorporationApplication anymore since it will be part of cooperative

Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts:100]: Doesn't a Cooperative IA schema include an IncorporationApplication block for the contents of the filing?
Could you link the schema?

Comment (vysakh-menon-aot) [reply src/mixins/filing-template-mixin.ts:100]: Yes cooperative is inside incorporationApplication. Please have a look into this PR https://github.com/bcgov/business-schemas/pull/69

File: src/interfaces/stepper-interfaces/CreateRules/rules-doc-interface.ts
Comment (severinbeauvais) [reply src/interfaces/stepper-interfaces/CreateRules/rules-doc-interface.ts]: Gotcha.

Since times and dates are overwhelmingly strings in JS, please add a line comment indicating what this is or where it comes from, so a future developer will know it's different from other times and dates. Thanks.

File: src/mixins/filing-template-mixin.ts
Comment (cameron-freshworks) [reply src/mixins/filing-template-mixin.ts:100]: Oh i see, its nested within `incorporationApplication. cooperative`. I got confused by your initial comment, i misunderstood.

File: src/components/CreateRules/UploadRules.vue
Comment (severinbeauvais) [inline src/components/CreateRules/UploadRules.vue:159]: Unused prop (isSummary).

File: src/components/CreateRules/UploadRulesSummary.vue
Comment (severinbeauvais) [reply src/components/CreateRules/UploadRulesSummary.vue:31]: (Vue)

File: src/components/CreateRules/UploadRulesSummary.vue
Comment (severinbeauvais) [inline src/components/CreateRules/UploadRulesSummary.vue:4]: Since this component is only used for summary purposes, you don't need isSummary. It complicates things unnecessarily.

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:100]: Argus, please update this code to nest `rules` and `memorandum` under `cooperative` on line 94 above. Or are these obsolete now that you use the file key and name?

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:97]: Argus, you probably need some code under line 161 to restore the values when a draft is reloaded.

File: src/resources/cooperative.ts
Comment (severinbeauvais) [inline src/resources/cooperative.ts:89]: I like it.

Please create an enum for 'link', 'text', etc.

File: src/store/mutations/mutations-model.ts
Comment (severinbeauvais) [inline src/store/mutations/mutations-model.ts:147]: I don't remember "rules confirmed" used anywhere. Did I miss it or is it unused?

File: src/mixins/filing-template-mixin.ts
Comment (argush3) [reply src/mixins/filing-template-mixin.ts:100]: So how does this work for everyone.  When building the filing, I will grab all properties that need to be saved and push it into the `cooperative` property for the IA filing.  i.e.

```
filing.filing.incorporationApplication.cooperative = {
          cooperativeAssociationType: this.getDefineCompanyStep.cooperativeType,
          rulesFileKey: this.getRules.docKey,
          rulesFileName: this.getRules.rulesDoc ? this.getRules.rulesDoc.name : null,
          rulesFileSize: this.getRules.rulesDoc ? this.getRules.rulesDoc.size : null,
          rulesFileLastModified: this.getRules.rulesDoc ? this.getRules.rulesDoc.lastModified : null,
          rulesConfirmed: this.getRules.rulesConfirmed
        }
```

When re-loading the draft back into the stores, I will populate `createRulesStep` in the state model with values from I from the `cooperative` property of the filing.

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:100]: Sound good, thanks!

PS is rules file name and size enough to re-populate the rules step when reloading a draft Coop IA?

File: src/store/mutations/mutations-model.ts
Comment (argush3) [reply src/store/mutations/mutations-model.ts:147]: `rulesConfirmed` is used in `UploadRules.vue`

```
<v-checkbox
  class="chk-rules"
  id="chk-confirm-rules"
  v-model="rulesConfirmed"
  label="I confirm the following items are included as required in the Rules of the Association:"
  @change="onRulesConfirmedChange()"
/>
```

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:97]: This comment is handled by another comment :)

File: src/store/mutations/mutations-model.ts
Comment (severinbeauvais) [reply src/store/mutations/mutations-model.ts:147]: Please confirm with BAs whether we can save that the rules were confirmed, or whether they should default to False every time we reload the draft (like the Certify checkbox).

Comment (severinbeauvais) [reply src/store/mutations/mutations-model.ts:147]: > `rulesConfirmed` is used in `UploadRules.vue`

Oh, I didn't see a getter for it.

Comment (argush3) [reply src/store/mutations/mutations-model.ts:147]: confirmed with Sienna that this will need to be reset on reload of draft.  will update to mimic what certify check box does.

File: src/mixins/filing-template-mixin.ts
Comment (argush3) [reply src/mixins/filing-template-mixin.ts]: updated using approach mentioned in my previous comment.

File: src/store/mutations/mutations-model.ts
Comment (argush3) [reply src/store/mutations/mutations-model.ts:147]: updated such that rulesConfirmed is reset each time draft filing is reloaded.

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts]: I'm good with the changes here. Up to Cam or Vysakh whether to resolve this conversation.

File: src/store/mutations/mutations-model.ts
Comment (severinbeauvais) [inline src/store/mutations/mutations-model.ts:142]: Use type "CreateRulesIF" for rules.

File: src/store/actions/actions-model.ts
Comment (severinbeauvais) [inline src/store/actions/actions-model.ts:130]: Use type "CreateRulesIF" for rules on line 129.

File: src/store/mutations/mutations-model.ts
Comment (severinbeauvais) [reply src/store/mutations/mutations-model.ts:147]: Oh I see... You have a getter and setter for the object (createRulesStep), and also individual setters for the object properties (rulesConfirmed, rulesDoc, docKey). I find that confusing.

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:11]: Use type "CreateRulesIF" for rules on line 38.

File: src/components/CreateRules/UploadRulesSummary.vue
Comment (argush3) [reply src/components/CreateRules/UploadRulesSummary.vue]: removed Vue import

File: src/components/ReviewConfirm/Summary.vue
Comment (argush3) [reply src/components/ReviewConfirm/Summary.vue]: removed `isSummary` prop

File: src/interfaces/stepper-interfaces/CreateRules/rules-doc-interface.ts
Comment (argush3) [reply src/interfaces/stepper-interfaces/CreateRules/rules-doc-interface.ts]: added comments to interface

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue:158]: removed `private readonly isSummary: boolean` line

File: src/components/CreateRules/UploadRulesSummary.vue
Comment (argush3) [reply src/components/CreateRules/UploadRulesSummary.vue]: removed `v-if="isSummary"`

File: src/components/CreateRules/UploadRules.vue
Comment (severinbeauvais) [reply src/components/CreateRules/UploadRules.vue:158]: Oops, you have to delete the decorator too.

File: src/resources/cooperative.ts
Comment (argush3) [reply src/resources/cooperative.ts]: created enum(`ItemTypes`) and updated all relevant references to use this new enum

File: src/store/mutations/mutations-model.ts
Comment (argush3) [reply src/store/mutations/mutations-model.ts]: updated to use `CreateRulesIF`

File: src/store/actions/actions-model.ts
Comment (argush3) [reply src/store/actions/actions-model.ts:130]: updated to use `CreateRulesIF`

File: src/mixins/filing-template-mixin.ts
Comment (argush3) [reply src/mixins/filing-template-mixin.ts:11]: updated to use `CreateRulesIF`

File: src/components/CreateRules/UploadRules.vue
Comment (argush3) [reply src/components/CreateRules/UploadRules.vue:158]: I think you are a machine 😄

Comment (argush3) [reply src/components/CreateRules/UploadRules.vue:156]: removed decorator too

File: src/store/mutations/mutations-model.ts
Comment (argush3) [reply src/store/mutations/mutations-model.ts:148]: will take another look and see if i can simplify things.  any specific component that does this well already i can reference?

File: src/components/CreateRules/UploadRules.vue
Comment (severinbeauvais) [reply src/components/CreateRules/UploadRules.vue:156]: Nah, I've made a lot of these kinds of mistakes before 😆

File: src/store/mutations/mutations-model.ts
Comment (severinbeauvais) [reply src/store/mutations/mutations-model.ts:148]: The closest I can think of is [this](https://github.com/bcgov/business-edit-ui/blob/1dfaa19b0c4257969c72b63abe69bddca6267458/src/views/Alteration.vue#L367), where there is only one getter and one setter, and whenever one of the fields needs to be updated, the whole object is updated.

In your case, I think you would have a few places to update to implement it this way.

Which do you think it less confusing?

Comment (argush3) [reply src/store/mutations/mutations-model.ts]: I think the approach [referenced](https://github.com/bcgov/business-edit-ui/blob/1dfaa19b0c4257969c72b63abe69bddca6267458/src/views/Alteration.vue#L367) is easier to understand.  I've the code to update the `rules` state model this way and also tried to minimize the variable usage among other clean up things.

**Review by severinbeauvais — APPROVED** [review]
Pre-approved for when you finish this.

**Review by vysakh-menon-aot — APPROVED** [review]

**Comment by argush3** [comment]
Merging into `main`.  Will take care of tests in separate PR.

**Pull request merged**
